### ISSUE TITLE
Bug: @typescript-eslint/no-redundant-type-constituents issue within declare module statement

### ISSUE THREAD

**Issue by ernestostifano** [typescript-eslint/typescript-eslint#11233]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

`./src/modules.d.ts`
```ts
declare module 'my-module' {
    import {EMyEnum} from './common.types.js';

    // INCORRECT: ESLINT HIGHLIGHTS AN ERROR
    function externalFunction(arg: EMyEnum | null): void;

    export {externalFunction};
}

```

```
ESLint: 'EMyEnum' is an 'error' type that acts as 'any' and overrides all other types in this union type. (@typescript-eslint/no-redundant-type-constituents)
```

It also seems to be an issue in TS:

`./src/index.ts`
```ts
import {externalFunction} from 'my-module';
import {internalFunction} from './utilities.js';

// INCORRECT: TS DOES NOT HIGHLIGHT THE ERROR
externalFunction('TEST');

// CORRECT: TS HIGHLIGHTS ERROR BECAUSE THE ARGUMENT MUST BE AN ENUM
internalFunction('TEST');
```

`./src/utilities.ts`
```ts
import {EMyEnum} from './common.types.js';

function internalFunction(arg: EMyEnum | null): void {
    console.log(arg);
}

export {internalFunction};

```

`./src/common.types.ts`
```ts
enum EMyEnum {
    Value = 'value'
}

export {EMyEnum};

```

### Reproduction Repository Link

https://github.com/ernestostifano/typescript-eslint-issue-1

### Repro Steps

1) Clone the repo.
2) See `README.md`.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `8.32.1` |
| `TypeScript`                           | `5.8.3`  |
| `ESLint`                               | `9.27.0` |
| `node`                                 | `22.6.0` |

**Comment by bradzacher** [comment]
I can't reproduce this in our playground
https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=CYUwxgNghgTiAEBbA9sArhBByRBPAtCuplvAN4BQ88AlogA7IwAu58Awsg8gHYg%2BsAvvABmMLvCxwoYZlgDcFKqLQ9ZNXvBAAPZiBg8oEAGKr1vABSwA5gC4OXRnwHwAPvB4YIASnsA3ZBpgRWUdRhY2HT0DI1M1Zg0eeEFFQQogA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNYmvAIYtaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxAGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

So this indicates that it's a problem with your setup - for whatever reason we can't find the types for your `common.types.js` file.

Which aligns with the error message you're seeing - it mentions the type is an "error" type because TS cannot resolve the types because it cannot find the file.

**Comment by ernestostifano** [comment]
@bradzacher thanks for your feedback. You can't replicate it because you are using a non-relative import within the declaration statement and the issue happens when doing a relative import (you cannot have multi-file environments in the playground).

But then, relative imports within module declaration scope are actually not allowed and that error was being suppressed by having `skipLibCheck` on.

Here you can find additional details: https://github.com/microsoft/TypeScript/issues/61744

I think we can close this issue.

Thanks!
